src/core/service.c

```diff
@@ -3093,6 +3093,8 @@ static void service_notify_message(Unit *u, pid_t pid, char **tags, FDSet *fds)
                         log_unit_warning(u, "Failed to parse MAINPID= field in notification message: %s", e);
                 else if (pid == s->control_pid)
                         log_unit_warning(u, "A control process cannot also be the main process");
+                else if (pid == getpid() || pid == 1)
```

Comment:
well, in the user case we also don't want to permit PID 1...

We should eventually add an even stronger check: that you cannot pass any foreign PIDs, that are not managed by the specific systemd instance. It's a bit sloppy that we don't check for that right now, but it's also not a high priority since it's not a security issue... the worst thing that might happen is that the user's own systemd --user instance gets a bit confused if a user app sends a weird sd_notify() msgs, but given that the sender is as privileged as the manager this is not problematic...

I didn't fix this properly yet, since the desired rules still aren't entirely clear to me (as we probably want to permit setting PIDs of other units as main PIDs of a service — which is done in the case of login/bash currently), hence I have only added that part to the TODO so far.

Anyway, the main reason I addded the explicit check for PID 1 for now is that we have assert(pid > 1) and assert(pid <= 1) at various place in the code ATM, and hence I'd at least want to validate our PIDs enough so that we don#t hit those...